pkg/proxy/endpointslicecache.go

```diff
@@ -0,0 +1,190 @@
+			if port.Name != nil {
+				svcPortName.Port = *port.Name
+			}
+			if _, exists := endpointInfoByServicePort[svcPortName]; !exists {
+				endpointInfoByServicePort[svcPortName] = map[string]Endpoint{}
+			}
+
+			// iterate through endpoints to add them to endpointInfoByServicePort
+			for i := range endpointSlice.Endpoints {
+				endpoint := &endpointSlice.Endpoints[i]
+				if len(endpoint.Addresses) < 1 {
+					klog.Warningf("ignoring invalid endpoint port %s with empty host", *port.Name)
+					continue
+				}
+
+				// Filter out the incorrect IP version case.
+				// Any endpoint port that contains incorrect IP version will be ignored.
+				if e.isIPv6Mode != nil && utilnet.IsIPv6String(endpoint.Addresses[0]) != *e.isIPv6Mode {
+					// Emit event on the corresponding service which had a different
+					// IP version than the endpoint.
+					utilproxy.LogAndEmitIncorrectIPVersionEvent(e.recorder, "endpointslice", endpoint.Addresses[0], serviceNN.Name, serviceNN.Namespace, "")
+					continue
+				}
+
+				isLocal := len(e.hostname) > 0 && endpoint.Topology["kubernetes.io/hostname"] == e.hostname
+				endpointInfo := newBaseEndpointInfo(endpoint.Addresses[0], int(*port.Port), isLocal)
+
+				// This logic ensures we're deduping potential overlapping endpoints
+				// isLocal should not vary between matching IPs, but if it does, we favor a true value here if it exists
+				if _, exists := endpointInfoByServicePort[svcPortName][endpointInfo.IP()]; !exists || isLocal {
+					if e.makeEndpointInfo != nil {
+						endpointInfoByServicePort[svcPortName][endpointInfo.IP()] = e.makeEndpointInfo(endpointInfo)
+					} else {
+						endpointInfoByServicePort[svcPortName][endpointInfo.IP()] = endpointInfo
+					}
+				}
+			}
+		}
+	}
+
+	// transform endpointInfoByServicePort into an endpointsMap with sorted IPs
+	for svcPortName, endpointInfoByIP := range endpointInfoByServicePort {
+		if len(endpointInfoByIP) > 0 {
+			endpointsMap[svcPortName] = []Endpoint{}
+			for _, endpointInfo := range endpointInfoByIP {
+				endpointsMap[svcPortName] = append(endpointsMap[svcPortName], endpointInfo)
+
+			}
+			// Ensure IPs are always returned in the same order to simplify diffing
+			sort.Sort(byIP(endpointsMap[svcPortName]))
+
+			if klog.V(3) {
+				newEPList := []string{}
+				for _, ep := range endpointsMap[svcPortName] {
+					newEPList = append(newEPList, ep.String())
+				}
+				klog.Infof("Setting endpoints for %q to %+v", svcPortName, newEPList)
+			}
+		}
+	}
+
+	return endpointsMap
+}
+
+// EndpointSliceCacheKeys returns cache keys used for a given EndpointSlice
+func EndpointSliceCacheKeys(endpointSlice *discovery.EndpointSlice) (types.NamespacedName, string) {
+	if len(endpointSlice.OwnerReferences) > 0 {
+		ownerRef := endpointSlice.OwnerReferences[0]
+		return types.NamespacedName{Namespace: endpointSlice.Namespace, Name: ownerRef.Name}, endpointSlice.Name
+	}
+	return types.NamespacedName{}, endpointSlice.Name
```

Comment:
Doesn't this collapse all slices without ownerrefs together? Don't we need to at least distinguish based on the namespace of the endpoint slice.

Reply:
Good catch. Added better logging + an early return in the calling function (cache.Update) that will help us avoid any invalid endpoint slices.